indexdata.go

```diff
@@ -388,6 +389,11 @@ func (d *indexData) iterateNgrams(query *query.Substring) (*ngramIterationResult
 
 	// Find the 2 least common ngrams from the string.
 	ngramOffs := splitNGrams([]byte(query.Pattern))
+	// PERF: Sort to increase the chances adjacent checks are in the same btree
+	// bucket (which can cause disk IO).
+	slices.SortFunc(ngramOffs, func(a, b runeNgramOff) bool {
```

Comment:
By sorting the ngrams, the slice of frequencies is now not in its natural order but reflects the sorted ngrams too. That means if we have several ngrams with the lowest frequency, before we picked the pair the was furthest apart, but with the sorting we lost that "spacial" property. Intuitively this should be rare?

Reply:
yeah I think I mentioned that on our call. It should be _super_ rare we have the same frequencies. But I can update the code to use the new offset param to tie break in this case just so we have the same behaviour as before.

Reply:
Alright doing this is a bit tricky so I tried to motivate why this optimization doesn't make sense. But then I realised the case where it does make sense. If your query string contains the same trigram twice maximing the distance is good for reducing candidate documents. eg a bad string like just `AAAAAAAAAAAA...AAA`. So I will make this work, the code just isn't as pleasant. I first implemented something generic that was nice to read, but took a big performance hit since go wasn't able to eliminate out of bounds checks.

Reply:
@stefanhengl I am gonna merge this as is and follow up with the PR to re-introduce this optimization. That way I can parallize dev and operational work :)